src/cc/bcc_btf.cc

```diff
@@ -21,10 +21,330 @@
+                .len = btf_ext->hdr->core_relo_len,
+                .min_rec_size = sizeof(struct bpf_core_relo),
+                .ext_info = &btf_ext->core_relo_info,
+                .desc = "core_relo",
+        };
+
+        return btf_ext_setup_info(btf_ext, &param);
+}
+
+static int btf_ext_parse_hdr(uint8_t *data, uint32_t data_size)
+{
+        const struct btf_ext_header *hdr = (struct btf_ext_header *)data;
+
+        if (data_size < offsetofend(struct btf_ext_header, hdr_len) ||
+            data_size < hdr->hdr_len) {
+                //pr_debug("BTF.ext header not found");
+                return -EINVAL;
+        }
+
+        if (hdr->magic == bswap_16(BTF_MAGIC)) {
+                //pr_warn("BTF.ext in non-native endianness is not supported\n");
+                return -ENOTSUP;
+        } else if (hdr->magic != BTF_MAGIC) {
+                //pr_debug("Invalid BTF.ext magic:%x\n", hdr->magic);
+                return -EINVAL;
+        }
+
+        if (hdr->version != BTF_VERSION) {
+                //pr_debug("Unsupported BTF.ext version:%u\n", hdr->version);
+                return -ENOTSUP;
+        }
+
+        if (hdr->flags) {
+                //pr_debug("Unsupported BTF.ext flags:%x\n", hdr->flags);
+                return -ENOTSUP;
+        }
+
+        if (data_size == hdr->hdr_len) {
+                //pr_debug("BTF.ext has no data\n");
+                return -EINVAL;
+        }
+
+        return 0;
+}
+
+void btf_ext__free(struct btf_ext *btf_ext)
+{
+	if((!btf_ext) || BCC_IS_ERR_VALUE((unsigned long)btf_ext))
+                return;
+        free(btf_ext->data);
+        free(btf_ext);
+}
+
+struct btf_ext *btf_ext__new(const uint8_t *data, uint32_t size)
+{
+        struct btf_ext *btf_ext;
+        int err;
+
+        btf_ext = (struct btf_ext*)calloc(1, sizeof(struct btf_ext));
+        if (!btf_ext)
+                return (struct btf_ext*)-ENOMEM;
+
+        btf_ext->data_size = size;
+        btf_ext->data = malloc(size);
+        if (!btf_ext->data) {
+                err = -ENOMEM;
+                goto done;
+        }
+        memcpy(btf_ext->data, data, size);
+
+        err = btf_ext_parse_hdr((uint8_t*)btf_ext->data, size);
+        if (err)
+                goto done;
+
+        if (btf_ext->hdr->hdr_len < offsetofend(struct btf_ext_header, line_info_len)) {
+                err = -EINVAL;
+                goto done;
+        }
+
+        err = btf_ext_setup_func_info(btf_ext);
+        if (err)
+                goto done;
+
+        err = btf_ext_setup_line_info(btf_ext);
+        if (err)
+                goto done;
+
+        if (btf_ext->hdr->hdr_len < offsetofend(struct btf_ext_header, core_relo_len)) {
+                err = -EINVAL;
+                goto done;
+        }
+
+        err = btf_ext_setup_core_relos(btf_ext);
+        if (err)
+                goto done;
+
+done:
+        if (err) {
+                btf_ext__free(btf_ext);
+                return (struct btf_ext*)err;
```

Comment:
err is 'int' type, converting to 'struct btf_ext *' will trigger the warning:
```
/home/yhs/work/bcc/src/cc/bcc_btf.cc: In function ‘btf_ext_vendored::btf_ext* btf_ext_vendored::btf_ext__new(const uint8_t*, uint32_t)’:                                            
/home/yhs/work/bcc/src/cc/bcc_btf.cc:276:41: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]                                                        
                 return (struct btf_ext*)err;                                                                                                                                       
                                         ^~~ 
```

You can do `(struct btf_ext *)(long)err`.

Reply:
fixed, but with a `uintptr_t` cast instead of `long`